Added bash script for auto-generating changelog and tagging new kubermatic releases

cbrgm opened the pull request:
**What this PR does / why we need it**:
Added bash script for auto-generating changelog and tagging new kubermatic releases.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1400

**Special notes for your reviewer**:

Requirements:
- `gchl` installed
- exported Github personal access token (GCHL_GITHUB_TOKEN)

The script does the following:
```
> Generate a changelog using gchl tool between HEAD and last tagged kubermatic release
> Append changes to existing CHANGELOG.md
> Commit new CHANGELOG.md
> Tag the created commit as a new kubermatic release
```

Usage: 
Execute the script in kubermatic root directory ./api/hack/changelog-gen.sh

```
# Create a changelog since last release, commit and create a new release tag
#
# Usage:
# changelog-gen.sh v2.x.x - create changelog, commit and tag new release, using closed PRs release-note
# changelog-gen.sh v2.x.x github-title - create changelog, commit and tag new release, using closed Githubs PRs title
```
P.S: https://github.com/kubermatic/gchl/pull/1 needs to be merged in order to auto execute `go get `  correctly if `gchl` is not installed.

I also increased the robustness of `gchl` using closed PRs `release-note` field. Now it can deal with all permutation of the word `NONE` e.g. `none`, `None`, `NoNe` and so on.

**Release note**:
```release-note
NONE
```

Commit:
Added changelog-gen.sh script

Commit:
Added changelog-gen.sh script

mrIncompetent commented:
@alvaroaleman Can you take care here?

alvaroaleman commented:
Will do

alvaroaleman commented on api/hack/changelog-gen.sh:
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
This script must be executable (`chmod +x`)

alvaroaleman commented on api/hack/changelog-gen.sh:
```diff
@@ -0,0 +1,54 @@
+if ! [ -x "$(command -v gchl)" ]; then
+	echo "Error: gchl not installed!" >&2
+	echo "Executing go get on https://github.com/kubermatic/gchl"
+	go get -v https://github.com/kubermatic/gchl
```
This will fail, go get must not be used in conjunction with a protocol prefix

alvaroaleman commented on api/hack/changelog-gen.sh:
```diff
@@ -0,0 +1,54 @@
+# Version flag is required for commit message and tag creation
+if [ "$1" == "" ]; then
+	echo "Please provide a version tag in format v0.0.0 -- [example: changelog-gen.sh v2.x.x]"
+	echo "Add release-notes flag to use closed PRs `release-notes` annotation content instead of PR title [example: changelog-gen.sh v2.x.x release-notes]"
```
You can not use backticks here, as this will result in the shell trying to execute the command `release-notes`

alvaroaleman commented on api/hack/changelog-gen.sh:
```diff
@@ -0,0 +1,54 @@
+CHANGELOG_FILE="CHANGELOG.md"
+
+gen_changelog() {
+	OUTPUT="$(gchl --for-version $1 since $(git tag -l | tail -n 1) $2)"; if [ $? -eq 1 ]; then exit 1; fi
```
It would be more helpful if this would use the second latest tag on the current branch, not the latest tag overall, e.G. via `git describe --tags $(git rev-list --tags --max-count=2)|tail -1`

alvaroaleman replied:
Actually the latest, in my testing I manually created a tag but I just realized the script does this which certainly makes sense

alvaroaleman reviewed (changes requested):
So I've tried executing this but it fails, even thought the commit in question definitely is in my branch:

```
$ ./api/hack/changelog-gen.sh v2.10.0
2018/07/26 10:38:48 Failed to find tag or branch name: v2.6.3-110-gd0abea3c
```

Any idea what the reason could be?

cbrgm commented:
Thanks for the review! I will do some rework and check the error you mentioned.

Commit:
Merge branch 'master' into changelog-gen

Commit:
Refactoring bash script

Commit:
Fix conflict

cbrgm commented:
To give you a short update on this one: I rewrote the bash script and fixed the points you mentioned @alvaroaleman . I'm still struggling with the tags. I seems like there is no way to display the human-readable names of a commit object like using the `git describe` command with the git library used in `gchl`.

The library only allows to list annotated and lightweight tags found in refs/tags namespace. So the library only allows me to list references like:

```
v1.0.1                            
v1.0 
```

but I'm unable to get a list like

```
v1.0.1                                              
v1.0-2-ga1050a44                                    
v1.0-1-g46b4c026                                    
v1.0 
```
Also the go-git `git describe` compatibility is not yet implemented. [See here](https://github.com/src-d/go-git/blob/master/COMPATIBILITY.md).
I'm still looking for a solution..

alvaroaleman commented:
@cbrgm Maybe just use `os/exec.Command("git", "my", "args").CombinedOutput()`? Would be nice to have it in golang but at the end of they day it would still shell out anyways, since many of those git subcommands are Perl scripts ;)

cbrgm commented:
Yeah, that was my intention to do. It felt a little bit strange using a library for git in go, but also using `os.Exec`.

might be interesting: [This](https://github.com/kubermatic/gchl/blob/master/pkg/git/git.go#L55) is the function, causing all the troubles. `g.repo.Tags()` behaves equal to `git tag -l`, also tried other approaches yesterday but it seems like using `os.Exec` is the only way to solve this issue here

Commit:
Added git rev-list command instead of git tag -l command

cbrgm commented:
Added the feature to compare hashes with branches/tags and other way around to `gchl`. See [PR](https://github.com/kubermatic/gchl/pull/2).

Also reworked the bash script. It's now using `git rev-list --tags --max-count=2)| tail -1` instead of `git tag -l`.

If you have some time the next days, feel free to try it out @alvaroaleman .

alvaroaleman committed:
Chmod +x

alvaroaleman committed:
Improve changelog generation

alvaroaleman committed:
Merge branch 'master' into changelog-gen

alvaroaleman commented:
@cbrgm I took the liberty of finishin this, PTAL

alvaroaleman commented:
This needs https://github.com/kubermatic/gchl/pull/2 to be merged, then IMHO its ready

alvaroaleman committed:
Simolify arg handling

cbrgm commented:
Thanks @alvaroaleman ! Looks good, well done.

cbrgm commented:
kubermatic/gchl#2 has been merged into master.

Commit:
Merge branch 'master' into changelog-gen

alvaroaleman committed:
Merge branch 'master' into changelog-gen
